File: ios/GuardApp/Core/GuardEngine/GuardEngine.swift
import Foundation

enum DeviceState: String, Equatable {
    case online
    case unstable
    case preAlert
    case alarming
    case paused
    case permissionLimited
    case lowPowerRisk
}

enum AlarmIntent: Equatable {
    case none
    case notifyAndVibrate(deviceId: String)
    case ringAndVibrate(deviceId: String)
}

struct GuardModeConfig: Equatable {
    let unstableAfter: TimeInterval
    let preAlertAfter: TimeInterval
    let strongAlertAfter: TimeInterval
    let allowsRinging: Bool

    static func config(for mode: GuardMode) -> GuardModeConfig {
        switch mode {
        case .outing:
            GuardModeConfig(unstableAfter: 6, preAlertAfter: 15, strongAlertAfter: 30, allowsRinging: true)
        case .indoor:
            GuardModeConfig(unstableAfter: 20, preAlertAfter: 50, strongAlertAfter: 100, allowsRinging: true)
        case .silent:
            GuardModeConfig(unstableAfter: 20, preAlertAfter: 50, strongAlertAfter: 100, allowsRinging: false)
        }
    }
}

struct DeviceObservation: Equatable {
    let deviceId: String
    let lastBleSeenAt: Date?
    let lastLanSeenAt: Date?
    let hasReliableBleHit: Bool
    let permissionBlocked: Bool
    let lowPowerRisk: Bool
    let pausedUntil: Date?

    var lastReliableSeenAt: Date? {
        let bleSeen = hasReliableBleHit ? lastBleSeenAt : nil
        return [bleSeen, lastLanSeenAt].compactMap { $0 }.max()
    }
}

struct GuardDecision: Equatable {
    let deviceId: String
    let state: DeviceState
    let alarmIntent: AlarmIntent
    let secondsSinceReliableSeen: TimeInterval?
}

final class GuardEngine {
    func evaluate(observation: DeviceObservation, mode: GuardMode, now: Date) -> GuardDecision {
        if let pausedUntil = observation.pausedUntil, pausedUntil > now {
            return GuardDecision(
                deviceId: observation.deviceId,
                state: .paused,
                alarmIntent: .none,
                secondsSinceReliableSeen: secondsSince(observation.lastReliableSeenAt, now: now)
            )
        }

        if observation.permissionBlocked {
            return GuardDecision(
                deviceId: observation.deviceId,
                state: .permissionLimited,
                alarmIntent: .none,
                secondsSinceReliableSeen: secondsSince(observation.lastReliableSeenAt, now: now)
            )
        }

        if observation.lowPowerRisk {
            return GuardDecision(
                deviceId: observation.deviceId,
                state: .lowPowerRisk,
                alarmIntent: .none,
                secondsSinceReliableSeen: secondsSince(observation.lastReliableSeenAt, now: now)
            )
        }

        guard let lastSeen = observation.lastReliableSeenAt else {
            return missingDecision(deviceId: observation.deviceId, elapsed: .infinity, mode: mode)
        }

        let elapsed = now.timeIntervalSince(lastSeen)
        return missingDecision(deviceId: observation.deviceId, elapsed: elapsed, mode: mode)
    }

    private func missingDecision(deviceId: String, elapsed: TimeInterval, mode: GuardMode) -> GuardDecision {
        let config = GuardModeConfig.config(for: mode)
        if elapsed < config.unstableAfter {
            return GuardDecision(deviceId: deviceId, state: .online, alarmIntent: .none, secondsSinceReliableSeen: elapsed)
        }
        if elapsed < config.preAlertAfter {
            return GuardDecision(deviceId: deviceId, state: .unstable, alarmIntent: .none, secondsSinceReliableSeen: elapsed)
        }
        if elapsed < config.strongAlertAfter || !config.allowsRinging {
            return GuardDecision(
                deviceId: deviceId,
                state: .preAlert,
                alarmIntent: .notifyAndVibrate(deviceId: deviceId),
                secondsSinceReliableSeen: elapsed
            )
        }
        return GuardDecision(
            deviceId: deviceId,
            state: .alarming,
            alarmIntent: .ringAndVibrate(deviceId: deviceId),
            secondsSinceReliableSeen: elapsed
        )
    }

    private func secondsSince(_ date: Date?, now: Date) -> TimeInterval? {
        date.map { now.timeIntervalSince($0) }
    }
}

File: android/app/src/main/java/com/deviceguard/core/guardengine/GuardEngine.kt
package com.deviceguard.core.guardengine

enum class GuardMode(val wireValue: String) {
    OUTING("outing"),
    INDOOR("indoor"),
    SILENT("silent");

    companion object {
        fun fromWireValue(value: String): GuardMode = entries.firstOrNull { it.wireValue == value }
            ?: throw IllegalArgumentException("Unsupported guard mode: $value")
    }
}

enum class DeviceState {
    ONLINE,
    UNSTABLE,
    PRE_ALERT,
    ALARMING,
    PAUSED,
    PERMISSION_LIMITED,
    LOW_POWER_RISK
}

sealed interface AlarmIntent {
    data object None : AlarmIntent
    data class NotifyAndVibrate(val deviceId: String) : AlarmIntent
    data class RingAndVibrate(val deviceId: String) : AlarmIntent
}

data class GuardModeConfig(
    val unstableAfterSeconds: Long,
    val preAlertAfterSeconds: Long,
    val strongAlertAfterSeconds: Long,
    val allowsRinging: Boolean
) {
    companion object {
        fun forMode(mode: GuardMode): GuardModeConfig = when (mode) {
            GuardMode.OUTING -> GuardModeConfig(
                unstableAfterSeconds = 6,
                preAlertAfterSeconds = 15,
                strongAlertAfterSeconds = 30,
                allowsRinging = true
            )
            GuardMode.INDOOR -> GuardModeConfig(
                unstableAfterSeconds = 20,
                preAlertAfterSeconds = 50,
                strongAlertAfterSeconds = 100,
                allowsRinging = true
            )
            GuardMode.SILENT -> GuardModeConfig(
                unstableAfterSeconds = 20,
                preAlertAfterSeconds = 50,
                strongAlertAfterSeconds = 100,
                allowsRinging = false
            )
        }
    }
}

data class DeviceObservation(
    val deviceId: String,
    val lastBleSeenAtEpochSeconds: Long?,
    val lastLanSeenAtEpochSeconds: Long?,
    val hasReliableBleHit: Boolean,
    val permissionBlocked: Boolean,
    val lowPowerRisk: Boolean,
    val pausedUntilEpochSeconds: Long?
) {
    val lastReliableSeenAtEpochSeconds: Long?
        get() = listOfNotNull(
            if (hasReliableBleHit) lastBleSeenAtEpochSeconds else null,
            lastLanSeenAtEpochSeconds
        ).maxOrNull()
}

data class GuardDecision(
    val deviceId: String,
    val state: DeviceState,
    val alarmIntent: AlarmIntent,
    val secondsSinceReliableSeen: Long?
)

class GuardEngine {
    fun evaluate(
        observation: DeviceObservation,
        mode: GuardMode,
        nowEpochSeconds: Long
    ): GuardDecision {
        val elapsed = observation.lastReliableSeenAtEpochSeconds?.let { nowEpochSeconds - it }

        if (observation.pausedUntilEpochSeconds != null && observation.pausedUntilEpochSeconds > nowEpochSeconds) {
            return GuardDecision(observation.deviceId, DeviceState.PAUSED, AlarmIntent.None, elapsed)
        }

        if (observation.permissionBlocked) {
            return GuardDecision(observation.deviceId, DeviceState.PERMISSION_LIMITED, AlarmIntent.None, elapsed)
        }

        if (observation.lowPowerRisk) {
            return GuardDecision(observation.deviceId, DeviceState.LOW_POWER_RISK, AlarmIntent.None, elapsed)
        }

        return missingDecision(
            deviceId = observation.deviceId,
            elapsedSeconds = elapsed ?: Long.MAX_VALUE,
            mode = mode
        )
    }

    private fun missingDecision(
        deviceId: String,
        elapsedSeconds: Long,
        mode: GuardMode
    ): GuardDecision {
        val config = GuardModeConfig.forMode(mode)
        return when {
            elapsedSeconds < config.unstableAfterSeconds -> GuardDecision(
                deviceId = deviceId,
                state = DeviceState.ONLINE,
                alarmIntent = AlarmIntent.None,
                secondsSinceReliableSeen = elapsedSeconds
            )
            elapsedSeconds < config.preAlertAfterSeconds -> GuardDecision(
                deviceId = deviceId,
                state = DeviceState.UNSTABLE,
                alarmIntent = AlarmIntent.None,
                secondsSinceReliableSeen = elapsedSeconds
            )
            elapsedSeconds < config.strongAlertAfterSeconds || !config.allowsRinging -> GuardDecision(
                deviceId = deviceId,
                state = DeviceState.PRE_ALERT,
                alarmIntent = AlarmIntent.NotifyAndVibrate(deviceId),
                secondsSinceReliableSeen = elapsedSeconds
            )
            else -> GuardDecision(
                deviceId = deviceId,
                state = DeviceState.ALARMING,
                alarmIntent = AlarmIntent.RingAndVibrate(deviceId),
                secondsSinceReliableSeen = elapsedSeconds
            )
        }
    }
}
